taskcat/amiupdater.py

```diff
@@ -0,0 +1,369 @@
+        # A word on this.
+        # - An instance of RegionalCodename is a representation of CODENAME and AMINAME.
+        # Since this module allows for multiple templates, *and* interrogates all templates at once,
+        # there's a risk I could end up with multiple instantations for each CODENAME/AMINAME combination.
+        # I maintain a dictionary of CODENAMEREGION -> Instance mappings, so if this class is instantated with the same
+        # arguments twice, the exact same memory pointer is returned.
+        try:
+            region = kwargs.get('region')
+            cn = kwargs.get('cn')
+        except KeyError:
+            raise
+        k = "{}{}".format(cn,region)
+        if k in cls._x.keys():
+            instance = cls._x.get(k)
+        else:
+            instance = super(Codenames, cls).__new__(cls)
+            cls._x[k] = instance
+        return instance
+
+    @classmethod
+    def objects(cls):
+        return [z for z in cls._x.values() if z.filters]
+
+    def __init__(self, cn, region, filters=None, *args, **kwargs):
+        self.region = region
+        self._results = None
+        self._filters = filters
+        super(RegionalCodename, self).__init__(cn, region, *args, **kwargs)
+
+class TemplateClass(object):
+    mapping_path = "Mappings/AWSAMIRegionMap"
+    metadata_path = "Metadata/AWSAMIRegionMap/Filters"
+    template_ext = ['.template', '.json', '.yaml', '.yml']
+    _regions = set()
+    _codenames = set()
+
+    @classmethod
+    def deep_get(cls, dictionary, keys, default=None):
+        zulu = reduce(lambda d, key: d.get(key, default) if isinstance(d, dict) else default, keys.split("/"), dictionary)
+        return zulu
+
+    @classmethod
+    def deep_set(cls, dictionary, keys, value):
+        for key in keys.split('/')[:-1]:
+            dic = dictionary.setdefault(key, {})
+        dictionary[keys[-1]] = value
+
+    @classmethod
+    def regions(cls):
+        return [x for x in list(cls._regions) if x is not 'AMI']
+
+    @classmethod
+    def _fetch_contents(cls, filename):
+        """Loads the template to inspect"""
+        with open(filename) as f:
+            tfdata = f.read()
+        stripped_tfdata = tfdata.strip()
+        if stripped_tfdata[0] in ['{', '['] and stripped_tfdata[-1] in ['}', ']']:
+            filetype = 'json'
+            loaded_template_data = json.loads(tfdata, object_pairs_hook=collections.OrderedDict)
+        else:
+            filetype = 'yaml'
+            loaded_template_data = cfy.ordered_safe_load(open(filename, 'rU'), object_pairs_hook=collections.OrderedDict)
+        return (filetype, loaded_template_data, tfdata)
+
+    def write(self):
+        #(qs-skldfjlskdf) // (AMI IDs updated by taskcat v0.8.23 180110/v1.2)
+        pass
+
+
+class TemplateObject(TemplateClass):
+    _objs = []
+    replacement_ami_map = {}
+
+    @classmethod
+    def objects(cls):
+        return cls._objs
+
+    def __init__(self, filename, all_regions=False):
```

Comment:
`__init__` method is pretty big, with several loops and conditions, may be more readable to split some of it out into private methods